src/include/rejson_api.h

```diff
@@ -21,52 +21,54 @@ typedef const void *RedisJSONKey;
 typedef const void *RedisJSON;
 
 typedef struct RedisJSONAPI_V1 {
-    /* RedisJSONKey functions */
-    RedisJSONKey (*openKey)(RedisModuleCtx *ctx, RedisModuleString *key_name);
-    RedisJSONKey (*openKeyFromStr)(RedisModuleCtx *ctx, const char *path);
-
-    void (*closeKey)(RedisJSONKey key);
-
-    /* RedisJSON functions
-     * Return NULL if path does not exist
-     * `count` can be NULL and return 0 for non array/object
-     **/
-    RedisJSON (*get)(RedisJSONKey key, const char *path, JSONType *type, size_t *count);
-
-    RedisJSON (*getAt)(RedisJSON jsonIn, size_t index, JSONType *type, size_t *count);
-
-    void (*close)(RedisJSON json);
-
-    /* RedisJSON value functions
-     * Return REDISMODULE_OK if RedisJSON is of the correct JSONType,
-     * else REDISMODULE_ERR is returned
-     **/
-    // Return int value from a Numeric, String or Bool field
-    int (*getInt)(RedisJSON json, long long *integer);
-    int (*getIntFromKey)(RedisJSONKey key, const char *path, long long *integer);
-
-    // Return double value from a Numeric, String or Bool field
-    int (*getDouble)(RedisJSON json, double *dbl);
-    int (*getDoubleFromKey)(RedisJSONKey key, const char *path, double *dbl);
-
-    // Return 0 or 1 as int value from a Numeric, String or Bool field
-    // Empty String returns 0
-    int (*getBoolean)(RedisJSON json, int *boolean);
-    int (*getBooleanFromKey)(RedisJSONKey key, const char *path, int *boolean);
-
-    // Callee has ownership of `str` - valid until api `close` is called
-    // Return the String or the String representation for Numeric or Bool
-    // FIXME: Handle null field
-    int (*getString)(RedisJSON json, const char **str, size_t *len);
-    int (*getStringFromKey)(RedisJSONKey key, const char *path, const char **str, size_t *len);
-
-
-    // Caller gains ownership of `str`
-    int (*getRedisModuleString)(RedisJSON json, RedisModuleString **str);
-    int (*getRedisModuleStringFromKey)(RedisJSONKey key, const char *path, RedisModuleString **str);
-
-    // Return 1 if type of key is JSON
-    int (*isJSON)(RedisModuleKey *redis_key);
+  /* RedisJSONKey functions */
+  RedisJSONKey (*openKey)(RedisModuleCtx *ctx, RedisModuleString *key_name);
+  RedisJSONKey (*openKeyFromStr)(RedisModuleCtx *ctx, const char *path);
+
+  void (*closeKey)(RedisJSONKey key);
+
+  /* RedisJSON functions
+   * Return NULL if path/index does not exist
+   **/
+  RedisJSON (*get)(RedisJSONKey key, const char *path, JSONType *type);
+  RedisJSON (*getAt)(RedisJSON json, size_t index, JSONType *type);
+
+  /* RedisJSON value functions
+   * Return REDISMODULE_OK if RedisJSON is of the correct JSONType,
+   * else REDISMODULE_ERR is returned
+   * */
+
+  // Return the length of Object/Array
+  int (*getLen)(RedisJSON json, size_t *count);
+
+  // Return the JSONType
+  int (*getType)(RedisJSON json, JSONType *type);
+
+  // Return int value from a Numeric field
+  int (*getInt)(RedisJSON json, long long *integer);
```

Comment:
If we will define that those function asserts that the type is correct we can make the API simpler and return the actual value (not via out parameters)

Reply:
@MeirShpilraien 
`getInt` may fail if path leads to a none numeric field - so panicking in this case seems too severe. WDYT?
`getType` - agreed in previous comment.

Reply:
I'd like to keep the error control on the client side, especially with read operations.

Reply:
@rafie OK, `getType` should not fail (only panics on null), and the rest of the APIs may return an error to the API client.